server/controllers/convo.js

```diff
@@ -0,0 +1,172 @@
+const Conversation = require('../models/Convo');
+
+module.exports = {
+  getAllConvos: asyncHandler(async (req, res) => {
+    let from = mongoose.Types.ObjectId(req.user.id);
+    Conversation.aggregate([{
+        $lookup: {
+          from: 'user',
+          localField: 'recipients',
+          foreignField: '_id',
+          as: 'recipientObj',
+        },
+      },
+     ])
+      .match({
+        recipients: {
+          $all: [{
+            $elemMatch: {
+              $eq: from
+            }
+          }]
+        }
+      })
+      .project({
+        'recipientObj.password': 0,
+        'recipientObj.__v': 0,
+      })
+      .exec(async (err, conversations) => {
+        if (err) {
+          console.log(err);
+          res.setHeader('Content-Type', 'application/json');
+          res.end(JSON.stringify({
+            message: 'Failure'
+          }));
+          res.sendStatus(500);
+        } else {
+          const populated = await Conversation.populate(conversations, {path: "recipients", select: '-__v -password -register_date'})
+          res.send(populated);
+        }
+      });
+  }),
+  getOneConvo: asyncHandler(async (req, res) => {
+    let user1 = mongoose.Types.ObjectId("60efa11241aa990dec93324e");
+    let user2 = mongoose.Types.ObjectId("60efa13a41aa990dec93324f");
+    Message.aggregate([{
+          $lookup: {
+            from: 'user',
+            localField: 'to',
+            foreignField: '_id',
+            as: 'toObj',
+          },
+        },
+        {
+          $lookup: {
+            from: 'user',
+            localField: 'from',
+            foreignField: '_id',
+            as: 'fromObj',
+          },
+        },
+      ])
+      .match({
+        $or: [{
+            $and: [{
+              to: user1
+            }, {
+              from: user2
+            }]
+          },
+          {
+            $and: [{
+              to: user2
+            }, {
+              from: user1
+            }]
+          },
+        ],
+      })
+      .project({
+        'toObj': 0,
+        'fromObj': 0,
+        "__v": 0
+      })
+      .exec((err, messages) => {
+        if (err) {
+          console.log(err);
+          res.setHeader('Content-Type', 'application/json');
+          res.end(JSON.stringify({
+            message: 'Failure'
+          }));
+          res.sendStatus(500);
+        } else {
+          res.send(messages);
+        }
+      });
+  }),
+  createMessage: asyncHandler(async (req, res) => {
+    let from = mongoose.Types.ObjectId(req.body.from);
+    let to = mongoose.Types.ObjectId(req.body.to);
+
+    Conversation.findOneAndUpdate({
+        recipients: {
+          $all: [{
+              $elemMatch: {
+                $eq: from
+              }
+            },
+            {
+              $elemMatch: {
+                $eq: to
+              }
+            },
+          ],
+        },
+      }, {
+        recipients: [req.body.from, req.body.to],
```

Comment:
We don't need to update the recipients here, as they are always going to be constant within a single conversation :)

Reply:
This is fixed.

Reply:
As a matter of fact, we do need this here because of how the conversation creation occurs. We look for a conversation to update, then check if there was no conversation found. If there was no conversation found, we create one and pass along the conversation id. So without this line of code, a new conversation will be created without this information.

Reply:
Ah I see! Thanks for the clarification.

Reply:
You got it